File: sude_py/sude/learning.py
from sklearn.neighbors import NearestNeighbors
from scipy.sparse import csr_matrix
from scipy.spatial.distance import cdist
from ._learning_utils import init_le_from_probability, sorted_distances_and_indices
from ._numba_kernels import (
    NUMBA_AVAILABLE,
    build_neighbor_probability_rows_sparse_snn,
    build_neighbor_probability_rows,
    fused_embedding_gradient,
    fused_embedding_gradient_2d_symmetric,
)
from .init_pca import init_pca
from .mds import mds
from .pca import pca
import math
import numpy as np
import time


def _adaptive_k2(n_samples):
    if n_samples < 9:
        return n_samples
    if n_samples > 1000:
        return int(np.ceil(np.log2(n_samples)) + 18)
    if n_samples > 50:
        return int(np.ceil(0.02 * n_samples)) + 8
    return 9


def _auto_block_size(n_samples, memory_budget_mb):
    if memory_budget_mb is None:
        return n_samples
    arrays_needed = 5
    bytes_per_float = np.dtype(np.float64).itemsize
    block_size = int(memory_budget_mb * 1024**2 / (n_samples * bytes_per_float * arrays_needed))
    return max(1, min(n_samples, block_size))


def _build_probability_matrix(
    X_samp,
    k1,
    get_knn,
    rnn,
    id_samp,
    no_dims,
    agg_coef,
    k2,
    memory_budget_mb=None,
):
    n_samples, n_features = X_samp.shape
    if k1 > 0:
        if NUMBA_AVAILABLE:
            landmark_knn = get_knn[id_samp]
            if memory_budget_mb is None:
                row, col, values = build_neighbor_probability_rows(
                    X_samp,
                    landmark_knn,
                    rnn,
                    k2,
                    agg_coef,
                )
            else:
                row, col, values = build_neighbor_probability_rows_sparse_snn(
                    X_samp,
                    landmark_knn,
                    rnn,
                    k2,
                    agg_coef,
                    block_rows=_auto_block_size(n_samples, memory_budget_mb),
                )
            return csr_matrix((values, (row, col)), shape=(n_samples, n_samples))

        row = []
        col = []
        values = []
        landmark_knn = get_knn[id_samp]
        knn_rnn_mat = rnn[landmark_knn]
        for i in range(n_samples):
            snn_id = np.isin(landmark_knn, landmark_knn[i]).astype(int)
            nn_id = np.where(np.max(snn_id, axis=1) == 1)[0]
            snn = np.zeros((1, n_samples))
            snn[:, nn_id] = np.sum(knn_rnn_mat[nn_id] * snn_id[nn_id], axis=1)
            mod_dis = (
                (1 - snn / max(np.max(snn), np.finfo(float).tiny)) ** agg_coef
                * cdist(X_samp[i : i + 1, :], X_samp)
            )
            sort_dis, idx = sorted_distances_and_indices(mod_dis)
            mean_samp_dis_squared = np.square(np.mean(sort_dis[0, :k2]))
            values.extend(
                np.exp(
                    -0.5
                    * np.square(sort_dis[0, :k2])
                    / np.maximum(mean_samp_dis_squared, np.finfo(float).tiny)
                )
            )
            row.extend((i * np.ones((k2, 1))).flatten().tolist())
            col.extend(idx[0, :k2])
        return csr_matrix((values, (row, col)), shape=(n_samples, n_samples))

    if n_samples > 5000 and n_features > 50:
        reduced = init_pca(X_samp, no_dims, 0.8)
        samp_dis, samp_knn = NearestNeighbors(n_neighbors=k2).fit(reduced).kneighbors(reduced)
    else:
        samp_dis, samp_knn = NearestNeighbors(n_neighbors=k2).fit(X_samp).kneighbors(X_samp)
    mean_samp_dis_squared = np.square(np.mean(samp_dis, axis=1))
    values = np.exp(
        -0.5
        * np.square(samp_dis)
        / np.maximum(mean_samp_dis_squared[:, np.newaxis], np.finfo(float).tiny)
    )
    return csr_matrix(
        (
            values.flatten(),
            ([i for i in range(n_samples) for _ in range(k2)], samp_knn.flatten()),
        ),
        shape=(n_samples, n_samples),
    )


def _initialize_embedding(X_samp, probability, no_dims, initialize):
    if initialize == "le":
        return init_le_from_probability(probability, no_dims)
    if initialize == "pca":
        return pca(X_samp, no_dims)
    if initialize == "mds":
        return mds(X_samp, no_dims)
    raise ValueError("initialize must be one of {'le', 'pca', 'mds'}")


def _dense_gradient(probability_dense, y, alpha, momentum, previous_gradient):
    distances = cdist(y, y) ** 2
    q1 = 1 / (1 + np.log(1 + distances))
    qq1 = 1 / (1 + distances)
    q = q1 / (np.sum(q1) - y.shape[0])
    pro_mat_y = 4 * (probability_dense - q) * q1 * qq1
    gradient = np.sum(pro_mat_y, axis=0)[:, np.newaxis] * y - pro_mat_y @ y
    return y - alpha * (gradient + momentum * previous_gradient), gradient


def _block_gradient(probability, y, alpha, momentum, previous_gradient, block_size):
    n_samples, no_dims = y.shape
    p_gradient = np.zeros((n_samples, no_dims))
    q_gradient = np.zeros((n_samples, no_dims))
    sum_q = 0.0
    for start in range(0, n_samples, block_size):
        stop = min(start + block_size, n_samples)
        idx = np.arange(start, stop)
        distances = cdist(y[idx], y) ** 2
        q1 = 1 / (1 + np.log(1 + distances))
        qq1 = 1 / (1 + distances)
        p_mat = -4 * probability[idx, :].multiply(q1).multiply(qq1).toarray()
        q_mat = -4 * q1**2 * qq1
        len_blk = idx.shape[0]
        id_pq = np.column_stack((np.arange(len_blk), start + np.arange(len_blk)))
        p_mat[id_pq[:, 0], id_pq[:, 1]] = p_mat[id_pq[:, 0], id_pq[:, 1]] - np.sum(
            p_mat,
            axis=1,
        )
        q_mat[id_pq[:, 0], id_pq[:, 1]] = q_mat[id_pq[:, 0], id_pq[:, 1]] - np.sum(
            q_mat,
            axis=1,
        )
        p_gradient[idx] = p_mat @ y
        q_gradient[idx] = q_mat @ y
        sum_q += np.sum(q1)
    gradient = p_gradient - q_gradient / (sum_q - n_samples)
    return y - alpha * (gradient + momentum * previous_gradient), gradient


def _fused_gradient(probability, y, alpha, momentum, previous_gradient):
    if y.shape[1] == 2:
        gradient = fused_embedding_gradient_2d_symmetric(probability, y)
    else:
        gradient = fused_embedding_gradient(probability, y)
    return y - alpha * (gradient + momentum * previous_gradient), gradient


def learning(
    X_samp,
    k1,
    get_knn,
    rnn,
    id_samp,
    no_dims,
    initialize,
    agg_coef,
    T_epoch,
    memory_budget_mb=None,
    return_profile=False,
):
    profile = {}
    total_start = time.perf_counter()
    n_samples, _ = X_samp.shape
    k2 = _adaptive_k2(n_samples)
    probability_start = time.perf_counter()
    probability = _build_probability_matrix(
        X_samp,
        k1,
        get_knn,
        rnn,
        id_samp,
        no_dims,
        agg_coef,
        k2,
        memory_budget_mb=memory_budget_mb,
    )
    probability = (probability + probability.transpose()) / 2
    profile["probability_seconds"] = time.perf_counter() - probability_start
    initialize_start = time.perf_counter()
    y = _initialize_embedding(X_samp, probability, no_dims, initialize)
    profile["initialize_seconds"] = time.perf_counter() - initialize_start
    normalize_start = time.perf_counter()
    probability = probability / (probability.sum() - n_samples)
    profile["normalize_probability_seconds"] = time.perf_counter() - normalize_start

    block_size = _auto_block_size(n_samples, memory_budget_mb)
    use_fused_gradient = NUMBA_AVAILABLE
    use_dense_gradient = (not use_fused_gradient) and block_size >= n_samples
    probability_dense = probability.toarray() if use_dense_gradient else None

    max_alpha = 2.5 * n_samples
    min_alpha = 2 * n_samples
    warm_step = 10
    previous_gradient = np.zeros((n_samples, no_dims))
    epoch = 1
    gradient_seconds = 0.0
    while epoch <= T_epoch:
        if epoch <= warm_step:
            alpha = max_alpha
        else:
            alpha = min_alpha + 0.5 * (max_alpha - min_alpha) * (
                1 + np.cos(np.pi * ((epoch - warm_step) / (T_epoch - warm_step)))
            )
        momentum = (epoch - 1) / (epoch + 2)
        gradient_start = time.perf_counter()
        if use_fused_gradient:
            y, gradient = _fused_gradient(probability, y, alpha, momentum, previous_gradient)
        elif use_dense_gradient:
            y, gradient = _dense_gradient(probability_dense, y, alpha, momentum, previous_gradient)
        else:
            y, gradient = _block_gradient(
                probability,
                y,
                alpha,
                momentum,
                previous_gradient,
                block_size,
            )
        gradient_seconds += time.perf_counter() - gradient_start
        previous_gradient = gradient
        epoch += 1

    profile["gradient_seconds"] = gradient_seconds
    profile["total_seconds"] = time.perf_counter() - total_start
    profile["n_samples"] = n_samples
    profile["k2"] = k2
    profile["used_fused_gradient"] = bool(use_fused_gradient)
    print(str(epoch - 1) + " epochs have been computed!")
    if return_profile:
        return y, k2, profile
    return y, k2


def memory_budget_for_large(n_samples):
    no_blocks = math.ceil(n_samples / 3000)
    block_size = math.ceil(n_samples / no_blocks)
    arrays_needed = 5
    bytes_per_float = np.dtype(np.float64).itemsize
    return block_size * n_samples * bytes_per_float * arrays_needed / 1024**2

File: sude_mat/learning.m
function [Y, k2] = learning (X_samp, k1, get_knn, rnn, id_samp, no_dims, initialize, agg_coef, T_epoch, varargin)
% Sparse-P landmark learning with two execution modes:
% dense keeps the original learning_s vectorized gradient path for speed;
% block keeps the original learning_l memory-bounded gradient path.

paramNames = {'Mode','BlockSize','SNNBlockSize','MemoryBudgetMB'};
defaults = {'dense',[],[],4096};
[mode, block_size, snn_block_size, memory_budget_mb] = internal.stats.parseArgs(paramNames, defaults, varargin{:});
mode = validatestring(mode, {'dense','block'}, mfilename, 'Mode');

[N, ~] = size(X_samp);
k2 = adaptive_landmark_neighbors(N);

if strcmpi(mode, 'dense')
    P = build_sparse_probability_matrix_dense(X_samp, k1, get_knn, rnn, id_samp, no_dims, agg_coef, k2);
else
    if isempty(snn_block_size)
        snn_block_size = memory_safe_snn_block_size(N, memory_budget_mb);
    end
    snn_block_size = max(1, min(N, snn_block_size));
    P = build_sparse_probability_matrix_block(X_samp, k1, get_knn, rnn, id_samp, no_dims, agg_coef, k2, snn_block_size);
end
P = sparse((P + P'))/2;
clear get_knn rnn id_samp

Y = initialize_embedding(P, X_samp, no_dims, initialize, mode);

P = P/(sum(P(:))-N);

if strcmpi(mode, 'dense')
    Y = optimize_dense(P, Y, T_epoch);
else
    if isempty(block_size)
        block_size = memory_safe_block_size(N);
    end
    block_size = max(1, min(N, block_size));
    Y = optimize_block(P, Y, T_epoch, block_size);
end
disp([num2str(T_epoch),' epochs have been computed!']);

end

function k2 = adaptive_landmark_neighbors(N)
if (N < 9)
    k2 = N;
elseif(N > 1000)
    k2 = ceil(log(N)/log(2)) + 18;
elseif(N > 50)
    k2 = ceil(0.02*N) + 8;
else
    k2 = 9;
end
end

function P = build_sparse_probability_matrix_dense(X_samp, k1, get_knn, rnn, id_samp, no_dims, agg_coef, k2)
[N, dim] = size(X_samp);
if(k1 > 0)
    SNN = zeros(N, N);
    G = get_knn(id_samp,:);
    knn_rnn_mat = rnn(G);
    for i = 1:N
        snn_id = ismember(G,G(i,:));
        nn_id = find(max(snn_id,[],2)==1);
        SNN(i,nn_id) = sum(knn_rnn_mat(nn_id,:).*snn_id(nn_id,:),2);
        SNN(i,:) = SNN(i,:)./max(max(SNN(i,:)),realmin);
    end
    Dis = (1-SNN).^agg_coef.*pdist2(X_samp,X_samp);
    [sort_dis, idx] = sort(Dis, 2);
    row = zeros(N*k2,1);
    col = zeros(N*k2,1);
    p_val = zeros(N*k2,1);
    for i = 1:N
        pos = (i-1)*k2+1:i*k2;
        row(pos) = i;
        col(pos) = idx(i,1:k2);
        p_val(pos) = exp(-0.5*sort_dis(i,1:k2).^2./max(mean(sort_dis(i,1:k2)).^2,realmin));
    end
    P = sparse(row,col,p_val,N,N);
else
    if(N >= 5000 && dim >= 50)
        xx = init_pca(X_samp, no_dims, 0.8);
        [samp_knn, samp_dis]= knnsearch(xx,xx,'k',k2);
    else
        [samp_knn, samp_dis]= knnsearch(X_samp,X_samp,'k',k2);
    end
    P = sparse(repmat((1:N)',k2,1),samp_knn(:),exp(-0.5*samp_dis.^2./max(mean(samp_dis,2).^2,realmin)),N,N);
end
end

function P = build_sparse_probability_matrix_block(X_samp, k1, get_knn, rnn, id_samp, no_dims, agg_coef, k2, snn_block_size)
[N, dim] = size(X_samp);
if(k1 > 0)
    row = zeros(N*k2,1);
    col = zeros(N*k2,1);
    p_val = zeros(N*k2,1);
    G = get_knn(id_samp,:);
    knn_rnn_mat = rnn(G);
    for start_row = 1:snn_block_size:N
        stop_row = min(start_row+snn_block_size-1,N);
        [row_block, col_block, p_block] = build_sparse_probability_matrix_block_rows( ...
            X_samp, G, knn_rnn_mat, agg_coef, k2, start_row, stop_row);
        pos = (start_row-1)*k2+1:stop_row*k2;
        row(pos) = row_block;
        col(pos) = col_block;
        p_val(pos) = p_block;
    end
    P = sparse(row,col,p_val,N,N);
else
    if(N >= 5000 && dim >= 50)
        xx = init_pca(X_samp, no_dims, 0.8);
        [samp_knn, samp_dis]= knnsearch(xx,xx,'k',k2);
    else
        [samp_knn, samp_dis]= knnsearch(X_samp,X_samp,'k',k2);
    end
    P = sparse(repmat((1:N)',k2,1),samp_knn(:),exp(-0.5*samp_dis.^2./max(mean(samp_dis,2).^2,realmin)),N,N);
end
end

function [row_block, col_block, p_block] = build_sparse_probability_matrix_block_rows(X_samp, G, knn_rnn_mat, agg_coef, k2, start_row, stop_row)
N = size(X_samp,1);
block_len = stop_row - start_row + 1;
row_block = zeros(block_len*k2,1);
col_block = zeros(block_len*k2,1);
p_block = zeros(block_len*k2,1);
D_block = pdist2(X_samp(start_row:stop_row,:),X_samp);
for local_i = 1:block_len
    i = start_row + local_i - 1;
    snn_id = ismember(G,G(i,:));
    nn_id = find(max(snn_id,[],2)==1);
    snn = zeros(1,N);
    snn(nn_id) = sum(knn_rnn_mat(nn_id,:).*snn_id(nn_id,:),2);
    [sort_dis, idx] = sort((1-snn./max(max(snn),realmin)).^agg_coef.*D_block(local_i,:),2);
    local_pos = (local_i-1)*k2+1:local_i*k2;
    row_block(local_pos) = i;
    col_block(local_pos) = idx(1:k2);
    p_block(local_pos) = exp(-0.5*sort_dis(1:k2).^2./max(mean(sort_dis(1:k2)).^2,realmin));
end
end

function Y = initialize_embedding(P, X_samp, no_dims, initialize, mode)
if strcmp(initialize,'le')
    if strcmpi(mode, 'block')
        degree = full(sum(P,1))';
        Dg = spdiags(degree, 0, size(P,1), size(P,2));
        Dg_sqrt = spdiags(sqrt(degree), 0, size(P,1), size(P,2));
        L = Dg_sqrt*(Dg - P)*Dg_sqrt;
    else
        Dg = diag(sum(P));
        L = Dg - P;
        L = Dg.^(0.5)*L*Dg.^(0.5);
    end
    [Y, ~] = eigs(L,no_dims+1,'smallestabs');
    Y(:,1)=[];
elseif strcmp(initialize,'pca')
    Y = pca(X_samp,no_dims);
elseif strcmp(initialize,'mds')
    Y = mds(X_samp,no_dims);
end
end

function Y = optimize_dense(P, Y, T_epoch)
[N, no_dims] = size(Y);
max_alpha = 2.5*N;
min_alpha = 2*N;
warm_step = 10;
preGrad = zeros(N,no_dims);
epoch = 1;
while epoch <= T_epoch
    if(epoch <= warm_step)
        alpha = max_alpha;
    else
        alpha = min_alpha + 0.5*(max_alpha-min_alpha)*(1+cos(pi*((epoch-warm_step)/(T_epoch-warm_step))));
    end
    D = pdist2(Y,Y).^2;
    Q1 = 1./(1+log(1+D));
    QQ1 = 1./(1+D);
    Q = Q1/(sum(Q1(:))-N);
    ProMatY = 4*(P-Q).*Q1.*QQ1;
    grad = (diag(sum(ProMatY))-ProMatY)*Y;
    Y = Y - alpha*(grad+(epoch-1)./(epoch+2)*preGrad);
    preGrad = grad;
    epoch = epoch + 1;
end
end

function Y = optimize_block(P, Y, T_epoch, block_size)
[N, no_dims] = size(Y);
no_blocks = ceil(N/block_size);
mark = zeros(no_blocks,2);
for i=1:no_blocks
   mark(i,:) = [(i-1)*block_size+1,min(i*block_size,N)];
end

max_alpha = 2.5*N;
min_alpha = 2*N;
warm_step = 10;
preGrad = zeros(N,no_dims);
epoch = 1;
while epoch <= T_epoch
    if(epoch <= warm_step)
        alpha = max_alpha;
    else
        alpha = min_alpha + 0.5*(max_alpha-min_alpha)*(1+cos(pi*((epoch-warm_step)/(T_epoch-warm_step))));
    end
    Pgrad = zeros(N,no_dims);
    Qgrad = zeros(N,no_dims);
    sumQ = 0;
    for i = 1:no_blocks
        idx = mark(i,1):mark(i,2);
        D_block = pdist2(Y(idx,:),Y).^2;
        Q1_block = 1./(1+log(1+D_block));
        QQ1_block = 1./(1+D_block);
        Pmat = -4*P(idx,:).*Q1_block.*QQ1_block;
        Qmat = -4*Q1_block.^2.*QQ1_block;
        len_blk = mark(i,2) - mark(i,1) + 1;
        idPQ = (len_blk*(mark(i,1)-1)+1):(len_blk+1):(len_blk*mark(i,2));
        Pmat(idPQ) = Pmat(idPQ) - sum(Pmat,2)';
        Qmat(idPQ) = Qmat(idPQ) - sum(Qmat,2)';
        Pgrad(idx,:) = Pmat*Y;
        Qgrad(idx,:) = Qmat*Y;
        sumQ = sumQ + sum(Q1_block(:));
    end
    grad = Pgrad - Qgrad/(sumQ-N);
    Y = Y - alpha*(grad + (epoch-1)/(epoch+2)*preGrad);
    preGrad = grad;
    epoch = epoch + 1;
end
end

function block_size = memory_safe_block_size(N)
max_block_entries = 8e6;
block_size = max(1, floor(max_block_entries/N));
block_size = min(N, block_size);
end

function block_size = memory_safe_snn_block_size(N, memory_budget_mb)
if isempty(memory_budget_mb) || isinf(memory_budget_mb)
    max_block_entries = 8e6;
else
    max_block_entries = max(1e6, memory_budget_mb * 1024^2 / 8 / 3);
end
block_size = max(1, floor(max_block_entries/N));
block_size = min(N, block_size);
end
